main.py

```diff
@@ -1,144 +1,197 @@
+        else:
+            raise HTTPException(status_code=400, detail=f"Search not supported for column: {search_column}")
 
     # Sorting
-    if order == "asc":
+    if sort_order == "asc":
         query = query.order_by(asc(sort_by))
     else:
         query = query.order_by(desc(sort_by))
 
-    db_pokemon = query.all()
-    db.close()
+    # Pagination
+    offset = (page - 1) * limit
+    db_pokemon = query.offset(offset).limit(limit).all()
 
     if not db_pokemon:
         raise HTTPException(status_code=404, detail="No Pokémon found")
 
-    return db_pokemon
+    return {"results": db_pokemon}
 
-@app.delete("/pokemon/{number}", response_model=PokemonResponse)
-def delete_pokemon(number: int = Path(description="The number of the Pokémon to delete")):
-    db = SessionLocal()
+
+@app.delete("/pokemon/{number}", response_model=PokemonPostPatchPutOutputSchema)
+def delete_pokemon(number: int = Path(description="The number of the Pokémon to delete"), db: Session = Depends(get_db), role: str = Depends(role_required(["admin", "user"]))):
     db_pokemon = db.query(Pokemon).filter(Pokemon.number == number).first()
     if db_pokemon is None:
-        db.close()
         raise HTTPException(status_code=404, detail="Pokémon not found")
     db.delete(db_pokemon)
     db.commit()
-    db.close()
     return db_pokemon
 
-@app.post("/pokemon/", response_model=PokemonResponse)
-def create_pokemon(pokemon: PokemonCreate = Body(...)):
-    db = SessionLocal()
+@app.post("/pokemon/", response_model=PokemonPostPatchPutOutputSchema)
+def create_pokemon(pokemon: PokemonPostPutInputSchema = Body(...), db: Session = Depends(get_db), role: str = Depends(role_required(["admin", "user"]))):
     db_pokemon = db.query(Pokemon).filter(Pokemon.number == pokemon.number, Pokemon.name == pokemon.name).first()
 
     if db_pokemon:
-        db.close()
         raise HTTPException(status_code=400, detail="Pokémon with this number and name already exists")
 
-    new_pokemon = Pokemon(
-        number=pokemon.number,
-        name=pokemon.name,
-        type_1=pokemon.type_1,
-        type_2=pokemon.type_2,
-        total=pokemon.total,
-        hp=pokemon.hp,
-        attack=pokemon.attack,
-        defense=pokemon.defense,
-        sp_atk=pokemon.sp_atk,
-        sp_def=pokemon.sp_def,
-        speed=pokemon.speed,
-        generation=pokemon.generation,
-        legendary=pokemon.legendary
-    )
+    new_pokemon = Pokemon(**pokemon.dict())
 
     db.add(new_pokemon)
     db.commit()
     db.refresh(new_pokemon)
-    db.close()
 
     return new_pokemon
 
 
-@app.put("/pokemon/{number}", response_model=PokemonResponse)
-def update_pokemon(number: int, pokemon: PokemonUpdate = Body(...)):
-    db = SessionLocal()
+@app.put("/pokemon/{number}", response_model=PokemonPostPatchPutOutputSchema)
+def update_pokemon(number: int, pokemon: PokemonPatchInputSchema = Body(...), db: Session = Depends(get_db), role: str = Depends(role_required(["admin", "user"]))):
```

Comment:
create variables for "admin", and "user" in constants.py and import them here
in case you want to change from "admin" to "administrator", you can update it in one place.